core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,1799 @@
+    case WitnessScriptKeyId.byte       => WitnessScriptKeyId
+    case BIP32DerivationPathKeyId.byte => BIP32DerivationPathKeyId
+    case _: Byte                       => UnknownKeyId
+  }
+
+  def fromBytes(key: ByteVector): PSBTOutputKeyId = {
+    if (key.isEmpty) {
+      UnknownKeyId
+    } else {
+      fromByte(key.head)
+    }
+  }
+
+  final case object RedeemScriptKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x00.byteValue
+  }
+
+  final case object WitnessScriptKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x01.byteValue
+  }
+
+  final case object BIP32DerivationPathKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x02.byteValue
+  }
+
+  final case object UnknownKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = Byte.MaxValue
+  }
+}
+
+sealed trait PSBTMap extends NetworkElement {
+  require(elements.map(_.key).groupBy(identity).values.forall(_.length == 1),
+          "All keys must be unique.")
+
+  def elements: Vector[PSBTRecord]
+
+  def bytes: ByteVector =
+    elements
+      .sortBy(_.key)
+      .foldLeft(ByteVector.empty)(_ ++ _.bytes) ++ hex"00"
+}
+
+case class GlobalPSBTMap(elements: Vector[GlobalPSBTRecord]) extends PSBTMap {
+
+  def getRecords[T <: GlobalPSBTRecord](key: PSBTGlobalKeyId): Vector[T] = {
+    elements
+      .filter(element => PSBTGlobalKeyId.fromByte(element.key.head) == key)
+      .asInstanceOf[Vector[T]]
+  }
+
+  /**
+    * Takes another GlobalPSBTMap and adds all records that are not contained in this GlobalPSBTMap
+    * @param other GlobalPSBTMap to be combined with
+    * @return A GlobalPSBTMap with the combined data of the two GlobalPSBTMaps
+    */
+  def combine(other: GlobalPSBTMap): GlobalPSBTMap = {
+    GlobalPSBTMap((this.elements ++ other.elements).distinct)
+  }
+}
+
+object GlobalPSBTMap {
+
+  def fromBytes(bytes: ByteVector): GlobalPSBTMap = {
+    @tailrec
+    def loop(
+        remainingBytes: ByteVector,
+        accum: Seq[GlobalPSBTRecord]): Vector[GlobalPSBTRecord] = {
+      if (remainingBytes.head == 0x00.byteValue) {
+        accum.toVector
+      } else {
+        val record = GlobalPSBTRecord.fromBytes(remainingBytes)
+        val next = remainingBytes.drop(record.bytes.size)
+
+        loop(next, accum :+ record)
+      }
+    }
+    GlobalPSBTMap(loop(bytes, Nil))
+  }
+}
+
+case class InputPSBTMap(elements: Vector[InputPSBTRecord]) extends PSBTMap {
+
+  def getRecords[T <: InputPSBTRecord](key: PSBTInputKeyId): Vector[T] = {
+    elements
+      .filter(element => PSBTInputKeyId.fromByte(element.key.head) == key)
+      .asInstanceOf[Vector[T]]
+  }
+
+  def isFinalized: Boolean =
+    getRecords(FinalizedScriptSigKeyId).nonEmpty || getRecords(
+      FinalizedScriptWitnessKeyId).nonEmpty
+
+  /** Finalizes this input if possible, returning None if not */
+  def finalize(input: TransactionInput): Option[InputPSBTMap] = {
+    if (isFinalized) {
+      Some(this)
```

Comment:
ditto above about throwing here, and turning this function into returning a `Try` rather than an `Option`

Reply:
@nkohen